polaris-core/src/main/java/org/apache/polaris/core/persistence/BasePersistence.java

```diff
@@ -0,0 +1,341 @@
+
+  /**
+   * Get change tracking versions for all specified entity ids.
+   *
+   * @param callCtx call context
+   * @param entityIds list of entity id
+   * @return list parallel to the input list of entity versions. If an entity cannot be found, the
+   *     corresponding element in the list will be null
+   */
+  @Nonnull
+  List<PolarisChangeTrackingVersions> lookupEntityVersions(
+      @Nonnull PolarisCallContext callCtx, List<PolarisEntityId> entityIds);
+
+  /**
+   * List all active entities of the specified type which are child entities of the specified parent
+   *
+   * @param callCtx call context
+   * @param catalogId catalog id for that entity, NULL_ID if the entity is top-level
+   * @param parentId id of the parent, can be the special 0 value representing the root entity
+   * @param entityType type of entities to list
+   * @return the list of entities_active records for the specified list operation
+   */
+  @Nonnull
+  List<PolarisEntityActiveRecord> listActiveEntities(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      @Nonnull PolarisEntityType entityType);
+
+  /**
+   * List active entities where some predicate returns true
+   *
+   * @param callCtx call context
+   * @param catalogId catalog id for that entity, NULL_ID if the entity is top-level
+   * @param parentId id of the parent, can be the special 0 value representing the root entity
+   * @param entityType type of entities to list
+   * @param entityFilter the filter to be applied to each entity. Only entities where the predicate
+   *     returns true are returned in the list
+   * @return the list of entities for which the predicate returns true
+   */
+  @Nonnull
+  List<PolarisEntityActiveRecord> listActiveEntities(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      @Nonnull PolarisEntityType entityType,
+      @Nonnull Predicate<PolarisBaseEntity> entityFilter);
+
+  /**
+   * List active entities where some predicate returns true and transform the entities with a
+   * function
+   *
+   * @param callCtx call context
+   * @param catalogId catalog id for that entity, NULL_ID if the entity is top-level
+   * @param parentId id of the parent, can be the special 0 value representing the root entity
+   * @param entityType type of entities to list
+   * @param limit the max number of items to return
+   * @param entityFilter the filter to be applied to each entity. Only entities where the predicate
+   *     returns true are returned in the list
+   * @param transformer the transformation function applied to the {@link PolarisBaseEntity} before
+   *     returning
+   * @return the list of entities for which the predicate returns true
+   */
+  @Nonnull
+  <T> List<T> listActiveEntities(
+      @Nonnull PolarisCallContext callCtx,
+      long catalogId,
+      long parentId,
+      @Nonnull PolarisEntityType entityType,
+      int limit,
```

Comment:
What's the use for `limit` when the caller cannot resume listing later?

Reply:
Looks like you can replace the last 3 parameters, if this returns a `Stream`

Reply:
For pagination we'd need a "proper" page token that guards against changes between any two pages, to make it impossible that the same entries are returned in more than one page.

For posterity, assume there are 5 tables:
* table-A
* table-C
* table-E
* table-G
* table-I

Assume a page-size of 2.

The first request yields
* table-A
* table-C

Then "table-B" is created

The second request **should** yield
* table-E
* table-G
* table-I

However, `org.apache.iceberg.rest.CatalogHandlers#paginate` uses integer indexes, so the second request would actually yield "table-C" again:
* table-C
* table-E
* table-G
* table-I

Implying that all `list*` calls return results in a deterministic order (e.g. ascending by name).

Reply:
See #273 for a pagination implementation that's not based on the integer index